File: first_time/DP/213.打家劫舍-ii.py
#
# @lc app=leetcode.cn id=213 lang=python3
#
# [213] 打家劫舍 II
#
from typing import List
# @lc code=start
class Solution:
    def rob(self, nums: List[int]) -> int:
        # 成环分为两种情况考虑
        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        
        # 不打劫第一家，考虑打劫最后一家
        result_1 = self.robRange(nums[1:])
        # 打劫第一家，不打劫最后一家
        result_2 = self.robRange(nums[:-1])
        return max(result_1, result_2)
    
    def robRange(self, nums):
        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        elif len(nums) == 2:
            return max(nums)

        dp = [0 for _ in range(len(nums))]
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, len(nums)):
            dp[i] = max(dp[i-2]+nums[i], dp[i-1])
        return dp[-1]

# @lc code=end

File: second_time/DP/213.打家劫舍-ii.cpp
/*
 * @lc app=leetcode.cn id=213 lang=cpp
 *
 * [213] 打家劫舍 II
 */
#include <vector>
#include <iostream>
using namespace std;

// @lc code=start
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 0) return 0;
        if (nums.size() == 1) return nums[0];
        int result1 = rob_single(nums, 0, nums.size()-2);
        int result2 = rob_single(nums, 1, nums.size()-1);
        return max(result1, result2);

    }


    int rob_single(vector<int>& nums, int start, int end){
        if (start==end) return nums[start];
        vector<int>dp(end-start+1, 0);
        dp[0] = nums[start];
        dp[1] = max(nums[start], nums[start+1]);
        for (int i = start+2; i <= end; i++){
            dp[i-start] = max(dp[i-1-start], dp[i-2-start]+nums[i]);
        }
        return dp[end-start];
    }

};
// @lc code=end

